sklearn/utils/openmp_helpers.pyx

```diff
@@ -0,0 +1,36 @@
+IF SKLEARN_OPENMP_SUPPORTED:
+    cimport openmp
+
+from joblib import effective_n_jobs
+
+
+cpdef _openmp_effective_n_threads(n_threads=None):
+    """Determine the effective number of threads used for parallel OpenMP calls
+
+    - For ``n_threads = None``, returns the minimum between
+      openmp.omp_get_max_threads() and joblib.effective_n_jobs(-1).
+    - For ``n_threads > 0``, use this as the maximal number of threads for
+      parallel OpenMP calls.
```

Comment:
Are we sure that we allow for over-subscription? Can't we raise a warning if `n_threads` is larger than `max_threads` and return `max_threads`?

Reply:
@ogrisel suggested so.

Reply:
I think it's fine to allow for oversubscription without warning as long as the default setting does not cause oversubscription. For some algorithms with a lot of IO a bit of oversubscription can actually be beneficial.